### DISCUSSION TITLE
How do I remove default Character commands - but also keep Admin commands

### DISCUSSION THREAD

**Discussion by RodRitter** [evennia/evennia#3424] [Evennia Questions] [answered]
Hey all,

I want to get rid of a lot of the default character commands and start from scratch. Things like 'look', 'get', 'say', etc.

My first solution involved commenting out the default commands, but I'm not sure changing the evennia package is wise.

Then I tried adding `CMDSET_CHARACTER=mycustomcommands.CharacterCmdSet` to my settings file, but it seems that removing `default_cmds.CharacterCmdSet` as the default, also removed some admin functionality.

Would creating a new Character class from Object (instead of `DefaultCharacter`) help, or would I lose a lot of functionality.

What is the best practice for overriding certain default commands?

Any help is appreciated.

**Comment by Griatch** [comment, marked as answer]
You never change the evennia core package. 

As outlined in the beginner tutorials, you make your changes in `mygame/commands/default_cmdsets.py`: 

```python
# in mygame/commands/default_cmdsets.py 

# if you have your own replacements in mygame/commands/mycommands.py
from commands import mycommands

class CharacterCmdSet(default_cmds.CharacterCmdSet):
    """
    The `CharacterCmdSet` contains general in-game commands like `look`,
    `get`, etc available on in-game Character objects. It is merged with
    the `AccountCmdSet` when an Account puppets a Character.
    """
 
    key = "DefaultCharacter"
 
    def at_cmdset_creation(self):
        """
        Populates the cmdset
        """
        super().at_cmdset_creation()
        #
        # any commands you add below will overload the default ones.
        #
        self.add(mycommands.CmdLook)        # Replace default `look` with one of your own (must have same key)
        self.remove(default_cmds.CmdGet)     # Revove default `get` completely
```

That is, you use `.add()` and `.remove()` to customize the default cmdset. There are other cmdsets for UnloggedIn, Accounts and Sessions that can be customized in the same way.

**Reply by RodRitter** [reply]
That's great, thanks for showing me. I understand how I should be doing it now

**Comment by cyberdexprojekt** [comment]
**I've tried removing or adding my own commands, but every time I get errors like this:**

_`Traceback (most recent call last):
  File "E:\Programy\python3118\Lib\site-packages\evennia\commands\cmdsethandler.py", line 180, in import_cmdset
    module = import_module(modpath, package="evennia")
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\Programy\python3118\Lib\importlib\__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "C:\Users\Ewela i Przemek\postapo\commands\default_cmdsets.py", line 18, in <module>
    from commands import mycommands
  File "C:\Users\Ewela i Przemek\postapo\commands\mycommands.py", line 1, in <module>
    COMMAND_DEFAULT_CLASS = utils.class_from_module(settings.COMMAND_DEFAULT_CLASS)
                            ^^^^^
NameError: name 'utils' is not defined

Compile/Run error when loading cmdset 'commands.default_cmdsets.CharacterCmdSet'.
(Traceback was logged 24-02-29 19:01:49+01)

Error encountered for cmdset at path 'commands.default_cmdsets.CharacterCmdSet'.
Replacing with fallback 'evennia.commands.default.cmdset_character.CharacterCmdSet'.`_

### **I have a mycommands.py file containing:**

_COMMAND_DEFAULT_CLASS = utils.class_from_module(settings.COMMAND_DEFAULT_CLASS)

class CmdHome(COMMAND_DEFAULT_CLASS):
    """
    move to your character's home location

    Usage:
      home

    Teleports you to your home location.
    """

    key = "home"
    locks = "cmd:perm(home) or perm(Builder)"
    arg_regex = r"$"

    def func(self):
        """Implement the command"""
        caller = self.caller
        home = caller.home
        if not home:
            caller.msg("You have no home!")
        elif home == caller.location:
            caller.msg("You are already home!")
        else:
            caller.msg("There's no place like home ...")
            caller.move_to(home, move_type="teleport")_

**### and in the default_cmdset.py file I added**

from commands import mycommands
...
        super().at_cmdset_creation()
        #
        # any commands you add below will overload the default ones.
        #
self.add(mycommands.CmdHome) 
self.remove(default_cmds.CmdHome)

### I would be grateful for a working instruction on how to add your own or remove existing commands. Or at least someone show their example mycommands.py and default_cmdsets.py files

**Reply by InspectorCaracal** [reply]
If you're adding your own version of a command, you don't need to remove the old one. Your command with the key of "home" will replace the default command just by adding it to your default command set there.

I recommend going through this tutorial page on how to actually create new working commands, since that seems to be the main thing that's causing your problems: https://www.evennia.com/docs/latest/Howtos/Beginner-Tutorial/Part1/Beginner-Tutorial-Adding-Commands.html

**Comment by cyberdexprojekt** [comment]
Thanks. I have read the documentation and now I know where I made mistakes. Now everything works.

**Reply by InspectorCaracal** [reply]
Glad to hear it! Good luck with the rest of your project!
